pallet/staking/src/lib.rs

```diff
@@ -735,215 +827,276 @@ pub mod pallet {
 
 								collator_v += nominator_v;
 
-								Some(IndividualExposure { who: n, vote: nominator_v })
+								Some(IndividualExposure { who: n.to_owned(), vote: nominator_v })
 							} else {
 								None
 							}
 						})
 						.collect();
 
-					((c, Exposure { vote: collator_v, nominators }), collator_v)
+					((c, Exposure { commission: cm, vote: collator_v, nominators }), collator_v)
 				})
 				.collect::<Vec<_>>();
 
 			collators.sort_by(|(_, a), (_, b)| b.cmp(a));
 
+			let cache_states = <ExposureCacheStates<T>>::get();
+
+			#[allow(deprecated)]
+			call_on_exposure!(cache_states, Next::remove_all(None))?;
+
 			collators
 				.into_iter()
 				.take(<CollatorCount<T>>::get() as _)
 				.map(|((c, e), _)| {
-					<NextExposures<T>>::insert(&c, e);
-
-					c
+					call_on_exposure!(cache_states, Next::insert(&c, e))
+						.map(|_| c)
+						.map_err(Into::into)
 				})
 				.collect()
 		}
 	}
 }
 pub use pallet::*;
 
-type RewardPoint = u32;
 type Power = u32;
 type Vote = u32;
 
 type DepositId<T> = <<T as Config>::Deposit as Stake>::Item;
 
-/// On session end handler.
-///
-/// Currently it is only used to control the inflation.
-pub trait OnSessionEnd<T>
+/// Inflation and reward manager.
+pub trait InflationManager<T>
 where
 	T: Config,
 {
 	/// Generic session termination procedures.
 	fn on_session_end() {
 		let inflation = Self::inflate();
 		let reward = Self::calculate_reward(inflation);
-		let actual_payout = <Pallet<T>>::payout(reward);
+		let actual_reward = <Pallet<T>>::dispatch_session_reward(reward);
 
-		Self::clean(inflation.unwrap_or(reward).saturating_sub(actual_payout));
+		if inflation != 0 {
+			Self::clear(inflation.saturating_sub(actual_reward));
+		}
 	}
 
 	/// Inflation settings.
-	fn inflate() -> Option<Balance> {
-		None
+	fn inflate() -> Balance {
+		0
 	}
 
 	/// Calculate the reward.
-	fn calculate_reward(maybe_inflation: Option<Balance>) -> Balance;
+	fn calculate_reward(inflation: Balance) -> Balance;
 
 	/// The reward function.
 	fn reward(who: &T::AccountId, amount: Balance) -> DispatchResult;
 
-	/// Clean the data; currently, only the unissued reward is present.
-	fn clean(_unissued: Balance) {}
+	/// Clear the remaining inflation.
+	fn clear(_remaining: Balance) {}
 }
-impl<T> OnSessionEnd<T> for ()
+impl<T> InflationManager<T> for ()
 where
 	T: Config,
 {
-	fn inflate() -> Option<Balance> {
-		None
-	}
-
-	fn calculate_reward(_maybe_inflation: Option<Balance>) -> Balance {
+	fn calculate_reward(_inflation: Balance) -> Balance {
 		0
 	}
 
 	fn reward(_who: &T::AccountId, _amount: Balance) -> DispatchResult {
 		Ok(())
 	}
+}
 
-	fn clean(_unissued: Balance) {}
+/// Exposure cache's state.
+#[allow(missing_docs)]
+#[cfg_attr(any(feature = "runtime-benchmarks", feature = "try-runtime"), derive(PartialEq))]
+#[derive(Encode, Decode, MaxEncodedLen, TypeInfo, RuntimeDebug)]
+pub enum ExposureCacheState {
+	Previous,
+	Current,
```

Comment:
It seems that `Current` is not used?

Reply:
Store current exposures to reward the current exposures in new session